File: lib/core/stats.dart
import 'calendar.dart';
import 'models.dart';
import 'store.dart';

class Stats {
  int totalAbsences = 0;
  int excused = 0;
  int unexcused = 0;
  int fullDays = 0;
  int passed = 0;
  int cancelled = 0;
  int consPresent = 0;

  /// ключ статуса -> количество
  final Map<String, int> byReason = {};

  /// предмет -> количество прогулов
  final Map<String, int> bySubject = {};

  /// предмет -> тип занятия -> количество
  final Map<String, Map<PairKind, int>> bySubjectKind = {};
}

/// День пропущен полностью?
///
/// Считаются только «зачётные» пары: без консультаций, без отменённых и без
/// «всё сдал». Если таких нет — день не пропущен. Если хоть на одной «пришёл» —
/// тоже нет. Если все пропущены (по любой причине) — день полный.
bool isFullDay(List<Mark> marks) {
  final countable = marks
      .where((m) => !m.consultation && !neutralKeys.contains(m.status))
      .toList();
  if (countable.isEmpty) return false;
  if (countable.any((m) => m.status == 'came')) return false;
  return countable.every((m) => absenceKeys.contains(m.status));
}

Stats computeStats(AppData data) {
  final s = Stats();
  data.marks.forEach((iso, marks) {
    for (final m in marks) {
      if (m.consultation) {
        if (m.status == 'present') s.consPresent++;
        continue;
      }
      if (absenceKeys.contains(m.status)) {
        s.totalAbsences++;
        s.byReason[m.status] = (s.byReason[m.status] ?? 0) + 1;
        s.bySubject[m.subject] = (s.bySubject[m.subject] ?? 0) + 1;
        final kinds = s.bySubjectKind.putIfAbsent(m.subject, () => {});
        kinds[m.kind] = (kinds[m.kind] ?? 0) + 1;
        if (excusedKeys.contains(m.status)) {
          s.excused++;
        } else if (unexcusedKeys.contains(m.status)) {
          s.unexcused++;
        }
      } else if (m.status == 'passed') {
        s.passed++;
      } else if (m.status == 'cancelled') {
        s.cancelled++;
      }
    }
    if (isFullDay(marks)) s.fullDays++;
  });
  return s;
}

class AbsenceRecord {
  final DateTime date;
  final String subject;
  final String status;

  const AbsenceRecord(this.date, this.subject, this.status);
}

List<AbsenceRecord> absenceRecords(AppData data) {
  final out = <AbsenceRecord>[];
  data.marks.forEach((iso, marks) {
    final day = DateTime.tryParse(iso);
    if (day == null) return;
    for (final m in marks) {
      if (m.consultation) continue;
      if (absenceKeys.contains(m.status)) {
        out.add(AbsenceRecord(day, m.subject, m.status));
      }
    }
  });
  out.sort((a, b) {
    final byDate = a.date.compareTo(b.date);
    return byDate != 0 ? byDate : a.subject.compareTo(b.subject);
  });
  return out;
}

/// Markdown-отчёт: сверху сводка, ниже подробности по каждому прогулу.
String buildReport(AppData data) {
  final s = computeStats(data);
  final records = absenceRecords(data);
  final b = StringBuffer()
    ..writeln('# Отчёт о прогулах — SkipKOB_Helper')
    ..writeln()
    ..writeln('_Сформировано: ${formatDateShort(DateTime.now())}_')
    ..writeln()
    ..writeln('## Краткая статистика')
    ..writeln()
    ..writeln('- Всего прогулов (пар): **${s.totalAbsences}**')
    ..writeln('  - по уважительной причине: **${s.excused}**')
    ..writeln('  - по неуважительной причине: **${s.unexcused}**')
    ..writeln('- Полностью пропущенных дней: **${s.fullDays}**')
    ..writeln('- Пар «всё сдал»: ${s.passed}')
    ..writeln('- Отменённых пар: ${s.cancelled}')
    ..writeln('- Посещено консультаций: ${s.consPresent}')
    ..writeln();

  if (s.byReason.isNotEmpty) {
    b
      ..writeln('### По причинам')
      ..writeln();
    final reasons = s.byReason.entries.toList()
      ..sort((x, y) => y.value.compareTo(x.value));
    for (final e in reasons) {
      b.writeln('- ${statusLabel(e.key)}: ${e.value}');
    }
    b.writeln();
  }

  if (s.bySubject.isNotEmpty) {
    b
      ..writeln('### По предметам')
      ..writeln();
    final subjects = s.bySubject.entries.toList()
      ..sort((x, y) => y.value.compareTo(x.value));
    for (final e in subjects) {
      b.writeln('- ${e.key}: ${e.value}');
    }
    b.writeln();
  }

  b
    ..writeln('## Подробности по каждому прогулу')
    ..writeln();
  if (records.isEmpty) {
    b.writeln('_Прогулов нет._');
  } else {
    b
      ..writeln('| Дата | Предмет | Причина |')
      ..writeln('| --- | --- | --- |');
    for (final r in records) {
      final wd = r.date.weekday <= 6 ? dayShortRu[dayKeys[r.date.weekday - 1]] : null;
      final dateStr =
          wd != null ? '${formatDateShort(r.date)} ($wd)' : formatDateShort(r.date);
      b.writeln('| $dateStr | ${r.subject} | ${statusLabel(r.status)} |');
    }
  }
  b.writeln();
  return b.toString();
}

File: SkipKOB_Helper_Classic/app/stats.py
"""Подсчёт статистики по сохранённым отметкам."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date

from .models import (
    ABSENCE_KEYS,
    EXCUSED_KEYS,
    NEUTRAL_KEYS,
    STATUS_BY_KEY,
    UNEXCUSED_KEYS,
)


def _is_cons(m: dict) -> bool:
    c = m.get("consultation")
    if c is None:
        c = str(m.get("subject", "")).startswith("#")
    return bool(c)


@dataclass
class Stats:
    total_absences: int = 0
    excused: int = 0
    unexcused: int = 0
    full_days: int = 0
    passed: int = 0
    cancelled: int = 0
    cons_present: int = 0
    cons_total: int = 0
    by_reason: dict[str, int] = field(default_factory=dict)   # ключ статуса -> кол-во
    by_subject: dict[str, int] = field(default_factory=dict)  # предмет -> кол-во прогулов
    by_subject_kind: dict[str, dict[str, int]] = field(default_factory=dict)  # предмет -> {тип -> кол-во}


def _is_full_day(marks: list[dict]) -> bool:
    """День полностью пропущен (см. правила в плане)."""
    countable = [
        m for m in marks
        if not _is_cons(m) and m["status"] not in NEUTRAL_KEYS
    ]
    if not countable:
        return False
    if any(m["status"] == "came" for m in countable):
        return False
    return all(m["status"] in ABSENCE_KEYS for m in countable)


def compute(data: dict) -> Stats:
    s = Stats()
    marks_by_date: dict[str, list[dict]] = data.get("marks", {})
    for iso, marks in marks_by_date.items():
        if not isinstance(marks, list):
            continue
        for m in marks:
            if not isinstance(m, dict):
                continue
            subject = m.get("subject", "")
            status = m.get("status", "")
            if _is_cons(m):
                s.cons_total += 1
                if status == "present":
                    s.cons_present += 1
                continue
            if status in ABSENCE_KEYS:
                s.total_absences += 1
                s.by_reason[status] = s.by_reason.get(status, 0) + 1
                s.by_subject[subject] = s.by_subject.get(subject, 0) + 1
                kind = m.get("kind", "lecture")
                bk = s.by_subject_kind.setdefault(subject, {})
                bk[kind] = bk.get(kind, 0) + 1
                if status in EXCUSED_KEYS:
                    s.excused += 1
                elif status in UNEXCUSED_KEYS:
                    s.unexcused += 1
            elif status == "passed":
                s.passed += 1
            elif status == "cancelled":
                s.cancelled += 1
        if _is_full_day(marks):
            s.full_days += 1
    return s


def absence_records(data: dict) -> list[tuple[date, str, str]]:
    """Список прогулов: (дата, предмет, ключ статуса), отсортированный по дате."""
    out: list[tuple[date, str, str]] = []
    for iso, marks in data.get("marks", {}).items():
        if not isinstance(marks, list):
            continue
        try:
            d = date.fromisoformat(iso)
        except ValueError:
            continue
        for m in marks:
            if not isinstance(m, dict):
                continue
            if _is_cons(m):
                continue
            if m.get("status") in ABSENCE_KEYS:
                out.append((d, m["subject"], m["status"]))
    out.sort(key=lambda r: (r[0], r[1]))
    return out


def reason_label(key: str) -> str:
    st = STATUS_BY_KEY.get(key)
    return st.label if st else key
